### DISCUSSION TITLE
client subscribe  Interval problem

### DISCUSSION THREAD

**Discussion by wushunqing** [eclipse-milo/milo#1135] [General]
Hi，@kevinherron, When I use Milo as the client，I used the method of subscribing to tagpoint。
I set samplingInterval less than or equal 10 second， Everything is normal, returning **one** data every 10 seconds，
 But greater than 10 seconds，Still returning **one** data in 10 seconds。 

       When I set this parameter **publishingInterval**, samplingInterval  and publishingInterval All set to 30 seconds, 
returning **three** data every 30seconds,I expect to only return one piece of data every **30** seconds，

How should I set???

![image](https://github.com/eclipse/milo/assets/20431475/23c8ddf1-28a7-4625-b1de-2fd748328a45)

**Comment by kevinherron** [comment]
Sorry, can you clarify, what are the values for each parameter, and what server are you connected to? And what behavior are you expecting?

**Reply by wushunqing** [reply]
Hi，@kevinherron, now, i set publishingInterval and samplingInterval  30000 ,I expect to only return **one** data every 30 seconds, but i got **three** data every 30。 the test env server is uacppserver, the prod env server is   wincc 7.5，The test results are the same.

![image](https://github.com/eclipse/milo/assets/20431475/c8eb6dfe-8800-4d64-8c4e-7c76cf0eee56)

**Reply by kevinherron** [reply]
Check to see if the server is revising the interval or if you are creating multiple items.

The client has no role in how many values the server returns, other than in the requested the parameters it creates the subscription and monitored items with. 

If the values aren't being revised you can look at Wireshark and see that the server is behaving incorrectly.

**Reply by wushunqing** [reply]
when I set  publishingInterval and samplingInterval less than or equal 10000, everything is ok, 
It will return a piece of data within the time you set,  but if greater than 10000，i got three data, why?

**Reply by wushunqing** [reply]
I think it may be related to this 10 second parameter

**Reply by kevinherron** [reply]
I don't know why. It sounds like an issue with the server. The client does not "poll" data. It simply receives data change notifications that the server sends.

**Reply by wushunqing** [reply]
What is the difference between publishingInterval and samplingInterval?

**Reply by kevinherron** [reply]
Sampling Interval is how often a Monitored Item should sample its underlying data.

Publishing Interval is how often a Subscription will gather up any queued values from all the Monitored Items that belong to it and then send a PublishResponse containing a notification message to the client.

It sounds like your expectations are correct - with Sampling Interval = 30s and Publish Interval = 30s, and only the 1 Monitored Item created, you should receive 1 value change every 30 seconds, not 3. But this behavior is the server's responsibility, not the client. All the client can do is request the desired parameters.

**Reply by wushunqing** [reply]
thanks

**Reply by wushunqing** [reply]
If I want to save only one piece of collected data every minute, how do I handle it from the client's perspective?

**Reply by kevinherron** [reply]
You should be able to set the sampling interval of the monitored item to 60s and a well behaved and compliant server would not deliver changes for that data point faster than every 60s.

**Reply by wushunqing** [reply]
This is really a problem with the server. After modifying the server, everything is normal
